lib/services/role.go

```diff
@@ -1543,41 +1547,52 @@ func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
 // CheckAccessToServer checks if a role has access to a node. Deny rules are
 // checked first then allow rules. Access to a node is determined by
 // namespaces, labels, and logins.
+//
+// Note, logging in this function only happens in debug mode, this is because
+// adding logging to this function (which is called on every server returned
+// by GetNodes) can slow down this function by 50x for large clusters!
 func (set RoleSet) CheckAccessToServer(login string, s Server) error {
 	var errs []error
 
-	// check deny rules first: a single matching namespace, label, or login from
+	// Check deny rules first: a single matching namespace, label, or login from
 	// the deny role set prohibits access.
 	for _, role := range set {
-		matchNamespace := MatchNamespace(role.GetNamespaces(Deny), s.GetNamespace())
-		matchLabels := MatchLabels(role.GetNodeLabels(Deny), s.GetAllLabels())
-		matchLogin := MatchLogin(role.GetLogins(Deny), login)
+		matchNamespace, namespaceError := MatchNamespace(role.GetNamespaces(Deny), s.GetNamespace())
+		matchLabels, labelError := MatchLabels(role.GetNodeLabels(Deny), s.GetAllLabels())
+		matchLogin, loginError := MatchLogin(role.GetLogins(Deny), login)
 		if matchNamespace && (matchLabels || matchLogin) {
-			errorMessage := fmt.Sprintf("role %v denied access to node %v: deny rule matched; match(namespace=%v, label=%v, login=%v)",
-				role.GetName(), s.GetHostname(), matchNamespace, matchLabels, matchLogin)
-			log.Warnf("[RBAC] Denied access to server: " + errorMessage)
-			return trace.AccessDenied(errorMessage)
+			if log.GetLevel() == log.DebugLevel {
```

Comment:
add comment here to explain why is it necessary

Reply:
Since this is done in three places, I added the comment to the function explaining why logging is behind a debug flag.